nemo/collections/llm/inference/base.py

```diff
@@ -0,0 +1,106 @@
+from pathlib import Path
+from typing import Optional
+
+import pytorch_lightning as pl
+import torch
+import torch.distributed
+from megatron.core.inference.common_inference_params import CommonInferenceParams
+from megatron.core.inference.engines.mcore_engine import MCoreEngine
+from megatron.core.inference.model_inference_wrappers.gpt.gpt_inference_wrapper import GPTInferenceWrapper
+from megatron.core.inference.model_inference_wrappers.inference_wrapper_config import InferenceWrapperConfig
+from megatron.core.inference.text_generation_controllers.simple_text_generation_controller import (
+    SimpleTextGenerationController,
+)
+from megatron.core.models.gpt.gpt_model import GPTModel as MCoreGPTModel
+from pytorch_lightning.trainer.states import TrainerFn
+
+import nemo.lightning as nl
+from nemo.lightning import io
+from nemo.lightning.pytorch.strategies.megatron_strategy import MegatronStrategy
+from nemo.lightning.pytorch.strategies.utils import RestoreConfig
+
+
+# We need this wrapper since mcore generate uses tokenizer.detokenize, tokenizer.tokenize to encode and decode prompts
+class MCoreTokenizerWrappper:
+    def __init__(self, tokenizer):
+        self.tokenizer = tokenizer
+        self.eod = tokenizer.eod
+        self.vocab_size = tokenizer.vocab_size
+
+    def detokenize(self, tokens):
+        return self.tokenizer.ids_to_text(tokens)
+
+    def tokenize(self, prompt):
+        return self.tokenizer.text_to_ids(prompt)
+
+
+# TODO: Move to lightning Fabric API.
+def _setup_trainer_and_restore_model(path: Path, trainer: nl.Trainer, model: pl.LightningModule):
+    assert isinstance(trainer.strategy, MegatronStrategy), "Only MegatronStrategy is supported for trainer.strategy."
+    assert trainer.strategy.context_parallel_size <= 1, "Context parallelism is not supported for inference."
```

Comment:
How about pipeline parallelism - is it supported for generation currently and, if yes, how?